executor/api/grpc/seldon/client.go

```diff
@@ -41,32 +47,70 @@ func NewSeldonGrpcClient(spec *v1.PredictorSpec, deploymentName string, annotati
 	smgc := SeldonMessageGrpcClient{
 		Log:            logf.Log.WithName("SeldonGrpcClient"),
 		callOptions:    opts,
-		conns:          make(map[string]*grpc.ClientConn),
+		conns:          make(map[string][]*grpc.ClientConn),
 		Predictor:      spec,
 		DeploymentName: deploymentName,
 		annotations:    annotations,
 	}
 	return &smgc
 }
 
+// TODO: Re-examine locks here, make sure concurrent reads/writes aren't overwriting each other
+// TODO: Verify this logic is necessary, or if we can just create numConns connections at once (simplifying logic)
+// TODO: Investigate a TLL for conns--may result in better load balancing if connections occasionally re-connect
 func (s *SeldonMessageGrpcClient) getConnection(host string, port int32, modelName string) (*grpc.ClientConn, error) {
+    s.RLock()
+    randNum := rand.Intn(numConns)
 	k := fmt.Sprintf("%s:%d", host, port)
-	if conn, ok := s.conns[k]; ok {
-		return conn, nil
+	if nodeConns, ok := s.conns[k]; ok {
+	    if c := nodeConns[randNum]; c != nil {
+	        defer s.RUnlock()
+	        return c, nil
+	    }
+	    s.RUnlock()
+
+	    c, err := s.createNewConn(modelName, host, port)
+	    if err != nil {
+	        return nil, err
+	    }
+
+	    s.Lock()
+	    s.conns[k][randNum] = c
+	    s.Unlock()
+
+		return s.conns[k][randNum], nil
 	} else {
-		opts := []grpc.DialOption{
-			grpc.WithInsecure(),
-		}
-		opts = append(opts, grpc2.AddClientInterceptors(s.Predictor, s.DeploymentName, modelName, s.annotations, s.Log))
-		conn, err := grpc.Dial(fmt.Sprintf("%s:%d", host, port), opts...)
-		if err != nil {
-			return nil, err
-		}
-		s.conns[k] = conn
-		return conn, nil
+	    s.RUnlock()
+	    connList := make([]*grpc.ClientConn, numConns)
+
+	    c, err := s.createNewConn(modelName, host, port)
+	    if err != nil {
+	        return nil, err
+	    }
+
+	    connList[randNum] = c
+
+	    s.Lock()
+	    s.conns[k] = connList
+	    s.Unlock()
+
+		return s.conns[k][randNum], nil
 	}
 }
```

Comment:
I've done this complicated stuff because it appears to better spread out the connections amongst the available processes on the server side. There's still no guarantee that each client will connect to a unique process though.